### DISCUSSION TITLE
Expected behavior for jax.distributed

### DISCUSSION THREAD

**Discussion by jjwilke** [jax-ml/jax#13439] [Q&A]
I am getting different behaviors for each of these scenarios:
1. 2 GPUs on a node in the same process
2. 1 GPU per process, 2 processes (recent commit: 7495a9e3705d67c498723240cfed3f14c5916567)
3. 1 GPU per process, 2 processes (0.3.25 release commit: a683186570404a05a2aea9f7e452400bb0299e57)

I launch the following simple program that adds two arrays and reduces the sum.
```
mpirun -n 2 python test.py
```

```
import jax
import os
import numpy as np
import jax.numpy as jnp
from mpi4py import MPI
from jax.experimental.pjit import pjit
from jax.experimental.pjit import PartitionSpec as P
from jax.experimental.maps import Mesh
from jax import make_jaxpr

rank = MPI.COMM_WORLD.Get_rank()
size = MPI.COMM_WORLD.Get_size()
rc = jax.distributed.initialize("localhost:1234", size, rank)

os.environ["CUDA_VISIBLE_DEVICES"] = str(rank)

def fxn(a,b):
    return (a + b).sum()

fp = pjit(fxn, in_axis_resources=(P('x',), P('x',)), out_axis_resources=None)

a = jnp.arange(8)
b = jnp.arange(8)
devices = np.array(jax.devices())
with Mesh(devices, ('x',)):
    if rank == 0:
        print(devices)
        print(make_jaxpr(fp)(a,b))
    c = fp(a,b)

if rank == 0:
    print(c)
```

For #2, I get the following output (jaxpr and result):

```
[StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0)
 StreamExecutorGpuDevice(id=1, process_index=1, slice_index=0)]
{ lambda ; a:i32[8] b:i32[8]. let
    c:i32[] = pjit[
      donated_invars=(False, False)
      in_positional_semantics=(<_PositionalSemantics.GLOBAL: 1>, <_PositionalSemantics.GLOBAL: 1>)
      in_shardings=(OpShardingSharding({devices=[2]0,1}), OpShardingSharding({devices=[2]0,1}))
      jaxpr={ lambda ; d:i32[8] e:i32[8]. let
          f:i32[8] = add d e
          g:i32[] = reduce_sum[axes=(0,)] f
        in (g,) }
      name=fxn
      out_positional_semantics=_PositionalSemantics.GLOBAL
      out_shardings=(OpShardingSharding({replicated}),)
      resource_env=ResourceEnv(Mesh(array([0, 1]), ('x',)), ())
    ] a b
  in (c,) }
56
```
#1 is basically the same except the GPUs are `process_index=0`

For #3, I get a different output and different jaxpr:
```
[StreamExecutorGpuDevice(id=0, process_index=0, slice_index=0)
 StreamExecutorGpuDevice(id=1, process_index=1, slice_index=0)]
{ lambda ; a:i32[8] b:i32[8]. let
    c:i32[] = pjit[
      donated_invars=(False, False)
      in_positional_semantics=(<_PositionalSemantics.LOCAL: 0>, <_PositionalSemantics.LOCAL: 0>)
      in_shardings=(OpShardingSharding({devices=[2]0,1}), OpShardingSharding({devices=[2]0,1}))
      jaxpr={ lambda ; d:i32[16] e:i32[16]. let
          f:i32[16] = add d e
          g:i32[] = reduce_sum[axes=(0,)] f
        in (g,) }
      name=fxn
      out_positional_semantics=_PositionalSemantics.LOCAL
      out_shardings=(OpShardingSharding({replicated}),)
      resource_env=ResourceEnv(Mesh(array([0, 1]), ('x',)), ())
    ] a b
  in (c,) }
112
```

So basically multi-GPU in one process and multi-process (recent commit) treat the input arrays `a` and `b` as global arrays of size 8 and create two shards of size 4 on each GPU.

Multi-process on the most recent release treats the input arrays `a` and `b` as global arrays of size 16 that have already been sharded. It was not obvious to me glancing through recent history why this behavior would have changed.

What is the expected or defined behavior? I would have thought the behavior would be the same regardless of whether multiple devices are in the same process or in different processes.  The "new" distributed behavior makes more sense to me since you would expected the answer to be independent of the number of GPUs or how the GPUs are distributed.

**Comment by hawkinsp** [comment]
@yashk2810

**Comment by yashk2810** [comment]
Please see this section related to jax.Array: https://jax.readthedocs.io/en/latest/jax_array_migration.html#breaking-change-for-pjit-after-switching-to-jax-array-for-host-local-inputs which introduces a breaking change to pjit.
